### DISCUSSION TITLE
Instructions for self-hosting?

### DISCUSSION THREAD

**Discussion by 0xabrar** [quirrel-dev/quirrel#954] [Q&A] [answered]
I'm using Vercel for my main website, and I decided to use a detached docker container running on a DigitalOcean droplet to try and get quirrel working. I got it working in localhost no problem, by using `QUIRREL_BASE_URL=@VERCEL_URL.` But the worker doesn't seem to be processing messages in prod. I'm not using CronJobs.

When I'm going to the Quirrel UI, for some reason I can never connect to the external host. Sending off an `enqueue` event, logging my docker container I get the following (highlighting specifically the error with invalid method:

```
{"level":10,"time":1644812593830,"pid":8,"hostname":"fe39f7b90e9a","err":{"type":"Error","message":"Parse Error: Invalid method encountered","stack":"Error: Parse Error: Invalid method encountered","bytesParsed":0,"code":"HPE_INVALID_METHOD","reason":"Invalid method encountered","rawPacket":{"type":"Buffer","data":[22,3,1,1,84,1,0,1,80,3,3,252,139,193,237,127,124,42,18,19,149,242,194,24,201,2,31,72,93,251,107,71,87,108,162,55,177,135,110,175,64,208,126,32,66,138,69,233,153,134,194,226,24,172,201,225,201,47,68,211,152,180,200,96,83,153,222,181,47,89,104,87,8,73,131,140,0,118,19,2,19,3,19,1,192,47,192,43,192,48,192,44,0,158,192,39,0,103,192,40,0,107,0,163,0,159,204,169,204,168,204,170,192,175,192,173,192,163,192,159,192,93,192,97,192,87,192,83,0,162,192,174,192,172,192,162,192,158,192,92,192,96,192,86,192,82,192,36,0,106,192,35,0,64,192,10,192,20,0,57,0,56,192,9,192,19,0,51,0,50,0,157,192,161,192,157,192,81,0,156,192,160,192,156,192,80,0,61,0,60,0,53,0,47,0,255,1,0,0,145,0,11,0,4,3,0,1,2,0,10,0,12,0,10,0,29,0,23,0,30,0,25,0,24,0,35,0,0,0,22,0,0,0,23,0,0,0,13,0,48,0,46,4,3,5,3,6,3,8,7,8,8,8,9,8,10,8,11,8,4,8,5,8,6,4,1,5,1,6,1,3,3,2,3,3,1,2,1,3,2,2,2,4,2,5,2,6,2,0,43,0,5,4,3,4,3,3,0,45,0,2,1,1,0,51,0,38,0,36,0,29,0,32,153,154,19,44,122,235,129,6,101,105,37,45,203,189,250,20,202,146,105,85,192,143,179,90,196,190,82,73,252,194,194,123]}},"msg":"client error"}
{"level":10,"time":1644812594066,"pid":8,"hostname":"fe39f7b90e9a","err":{"type":"Error","message":"Parse Error: Invalid method encountered","stack":"Error: Parse Error: Invalid method encountered","bytesParsed":0,"code":"HPE_INVALID_METHOD","reason":"Invalid method encountered","rawPacket":{"type":"Buffer","data":[22,3,1,1,84,1,0,1,80,3,3,204,127,17,177,165,148,192,251,147,212,9,218,112,171,24,148,103,41,139,105,224,115,109,30,16,146,17,196,88,240,62,121,32,137,149,142,103,149,219,28,92,64,250,218,14,65,142,134,105,13,144,250,20,134,232,23,50,190,162,168,186,215,38,225,76,0,118,19,2,19,3,19,1,192,47,192,43,192,48,192,44,0,158,192,39,0,103,192,40,0,107,0,163,0,159,204,169,204,168,204,170,192,175,192,173,192,163,192,159,192,93,192,97,192,87,192,83,0,162,192,174,192,172,192,162,192,158,192,92,192,96,192,86,192,82,192,36,0,106,192,35,0,64,192,10,192,20,0,57,0,56,192,9,192,19,0,51,0,50,0,157,192,161,192,157,192,81,0,156,192,160,192,156,192,80,0,61,0,60,0,53,0,47,0,255,1,0,0,145,0,11,0,4,3,0,1,2,0,10,0,12,0,10,0,29,0,23,0,30,0,25,0,24,0,35,0,0,0,22,0,0,0,23,0,0,0,13,0,48,0,46,4,3,5,3,6,3,8,7,8,8,8,9,8,10,8,11,8,4,8,5,8,6,4,1,5,1,6,1,3,3,2,3,3,1,2,1,3,2,2,2,4,2,5,2,6,2,0,43,0,5,4,3,4,3,3,0,45,0,2,1,1,0,51,0,38,0,36,0,29,0,32,98,65,127,37,75,247,253,109,131,124,26,38,106,41,222,245,134,225,232,128,35,208,7,114,237,69,74,51,120,221,200,4]}},"msg":"client error"}
{"level":10,"time":1644812594662,"pid":8,"hostname":"fe39f7b90e9a","module":"owl","msg":"Checking for jobs"}
{"level":10,"time":1644812594662,"pid":8,"hostname":"fe39f7b90e9a","module":"owl","msg":"Peeking into queue"}
{"level":10,"time":1644812594663,"pid":8,"hostname":"fe39f7b90e9a","module":"owl","result":["wait",21259519],"msg":"Checking for jobs finished"}
```
This is my `docker-compose.yml`, which I'm running with `docker-compose up -d`.

```
version: '3.7'
services:
  quirrel:
    image: ghcr.io/quirrel-dev/quirrel:main
    environment:
      REDIS_URL: redis://redis
      PASSPHRASES: (my-passphrase)
    ports:
      - '9181:9181'
    tty: true

  redis:
    image: redis
    volumes:
      - redis_data:/data
    tty: true

volumes:
  redis_data:
    external: false
```

I have all of my `.env` variables set, with the following:

```
QUIRREL_BASE_URL=mylinks.ai
QUIRREL_API_URL=(my-ip):9181
QUIRREL_TOKEN=(my-token)
QUIRREL_ENCRYPTION_SECRET=(my-secret)
```
I've tried tons of variations of different things at this point, so not really sure what I'm doing wrong. My guess is that something is going wrong with the secure message transfer.

**Comment by Skn0tt** [comment, marked as answer]
> My guess is that something is going wrong with the secure message transfer.

That was my first idea, as well. Quirrel's 9181 listens via HTTP. Since you specified `QUIRREL_API_URL` without a protocol, [it will default to `https`](https://github.com/quirrel-dev/quirrel/blob/ba995193d3db6f962c1e324f562f0a69db6daf04/src/client/config.ts#L36).

The hacky solution is to specify the protocol (`http://(my-ip):9181`). The right solution is to add a HTTPS-Proxy in front of Quirrel (super easy to do using something like [Caddy](https://caddyserver.com/docs/quick-starts/reverse-proxy)).

**Comment by 0xabrar** [comment]
Tried the `http` hack and that seemed to work out. Ended up getting `401` error though on the Vercel endpoint, found out all API endpoints use `same-origin` policy afterward. Know there's a way to make this work out with some middleware, but calling it quits after trying to get this setup for hours (ui.quirrel still doesn't work).

Found another solution that lets me relay requests with just an HTTP call in 5mins, went with that instead. 

Thanks for the help, wish I didn't just barely miss the hosted solution for Quirrel :(

**Reply by Skn0tt** [reply]
If `ui.quirrel.dev` doesn't work, you can try opening `(my-ip)` in the browser - it should be serving the Quirrel UI.
